File: 2013-06-07/python/exercise3.py
from pyplasm import *
import random

dom1D = INTERVALS(60)(16)
dominio = PROD([dom1D,dom1D])
DRAW = VIEW
mappaZ = {};

def montagne(punto):
	x = punto[0];
	y = punto[1];
	z =  (random.random()*3 * SIN(x) * COS(y))**2;
	ran = random.random()/5;
	return [x, y, ran+z];


def lagoF(max) :
	def l (punto):
		x = punto[0];
		y = punto[1];
		z = max;
		return [x, y, z];
	return l;	


dominioLago = INTERVALS(5)(10);
dominioLago = PROD([dominioLago,dominioLago]);
medio = 0.52;
lagoF = lagoF(medio);
lago = MAP(lagoF)(dominioLago);
lago = T([0,1])([-10,-10])(COLOR(BLUE)(lago))

territorio = MAP(montagne)(dominio);
territorio = T([1,2])([-30,-30])(territorio)

totale = STRUCT([lago,territorio])
#-  EX 3----------------------------------------------------------------------------------------------------
def bezierS1(controlpoints):
	return BEZIER(S1)(controlpoints)

def bezierS2(f):
	return BEZIER(S2)(f)

def mappedBezier_1D(controlpoints):
	x = bezierS1(controlpoints)
	return MAP(x)(INTERVALS(1)(32))

def mappedBezier_2D(functions):
	x = BEZIER(S2)(functions)
	return MAP(x)(dominio) 

def scalaRaggio(radius):
	def prova(punto):
		y = 5;
		return [punto[0]*radius,punto[1]*radius+y,punto[2]]
	return prova

def bezier_circumference(radius,z):
	controlpoints = [[1,0,z],[1,1,z],[0,1.62,z],[-1.22,1.22,z],[-2,0,z],[-1.22,-1.22,z],[0,-1.63,z],[1,-1,z],[1,0,z]]
	controlpoints = AA(scalaRaggio(radius))(controlpoints)
	return bezierS1(controlpoints)

def treeModel(raggio,z,altezzaTronco):
	circle = CIRCLE(raggio)([30,1])
	tronco = EXTRUDE([1,circle,altezzaTronco])
	baseCono = bezier_circumference(raggio+0.5,altezzaTronco);
	first = [0,0,altezzaTronco+1];
	second =[0,0,altezzaTronco]
	foglie = MAP(CONICALSURFACE([first,baseCono]))(dominio)
	baseFoglie = MAP(CONICALSURFACE([second,baseCono]))(dominio);
	return STRUCT([tronco,foglie,baseFoglie])

a = treeModel(1,0,20);
DRAW(a)
#a = bezier_circumference(10,0);
#DRAW(a)

File: 2013-06-07/javascript/exercise3.js
var dominio = DOMAIN([[-30, 30], [-30, 30]])([30,30]);

//Mappa per salvare la z di ogni punto
var mappaZ = {};

var montagneF = function(punto){
	var x = punto[0];
	var y = punto[1];
	var z =  Math.pow(Math.random()*3 * SIN(x) * COS(y),1.3);
	var random = Math.random()/5;
	z = random + z;
	mappaZ[punto[0]+","+punto[1]] = z;
	return [x, y, z];
}

var lagoF = function (max) {
	function l (punto) {
		var x = punto[0];
		var y = punto[1];
		var z = max;
		//console.log("["+ x + " , " + y + ", " +z + "]")
		return [x, y, z];
	}
	return l;	
}

var minimo = function(){
	var m = 100;
	for (var i = 24; i <= 28; i++) {
		for (var j = -5; j <= 5; j++) {
			if(mappaZ[i+","+j]<m) m=mappaZ[i+","+j];	
		};
	};
	return m;
}

var massimo = function(){
	var m = -100;
	for (var i = 24; i <= 28; i++) {
		for (var j = -5; j <= 5; j++) {
			if(mappaZ[[i+","+j]]>m){ 
				m=mappaZ[[i+","+j]]; 
			}
		};
	};
	return m;
}

//Ex1
var territorio = MAP(montagneF)(dominio);
DRAW(COLOR([160/255, 82/255, 45/255])(territorio));

//Ex 2
var min = minimo(mappaZ);
var max = massimo(mappaZ);
var dominioLago = DOMAIN([[0, 5], [-5, 5]])([1,1]);
var medio = 0.15;
lagoF = lagoF(medio);
var lago = MAP(lagoF)(dominioLago);
DRAW(COLOR([70/255,130/255,180/255])(lago))


//--------------------------------------------- ES3 ----------------------------------------
var domain2D = DOMAIN([[0,1],[0,1]])([30,30]);

function bezierS0(controlpoints){
  return BEZIER(S0)(controlpoints)
}

function bezierMappata_2D(functions){
  var x = BEZIER(S1)(functions)
  return MAP(x)(domain2D) 
}

function cerc(r,z){
  var points = [[1,0,0],[1,1,0],[0,1.7,0],[-1.7,1,0],[-1.7,0,0],[-1,-1.4,0],[0,-1.6,0],[1,-0.9,0],[1,0,0]];
  var c = points.map(function(point){return [point[0]*r,point[1]*r,point[2]+z]})
  var cerchio = bezierS0(c);
  return cerchio;
}

var treeModel = function(raggio,z,altezzaTronco){
	var baseAlbero = cerc(raggio,z);
	var fin = cerc(raggio,altezzaTronco);
	var tronco = bezierMappata_2D([fin,baseAlbero]);
	var baseCono = cerc(raggio+0.5,altezzaTronco);
	var first = [0,0,altezzaTronco+1];
	var second =[0,0,altezzaTronco]
	var foglie = MAP(CONICAL_SURFACE(first)(baseCono))(domain2D);
	var baseFoglie = MAP(CONICAL_SURFACE(second)(baseCono))(domain2D);
	return STRUCT([COLOR([150/255, 75/255, 0])(tronco),COLOR([0,1,0])(foglie),COLOR([0,1,0])(baseFoglie)])
}

a = treeModel(0.1,0,2);

for (var i = 20; i <25; i++) {
	for (var j = 10; j < 20; j++) {
		z = mappaZ[i+","+j];
		//si può fare anche scalato
		DRAW(T([0,1,2])([i+Math.random()*5,j+Math.random()*10,z])(a))
	};
};

for (var i = -10; i <-5; i++) {
	for (var j = 10; j < 15; j++) {
		z = mappaZ[i+","+j];
		//si può fare anche scalato
		DRAW(T([0,1,2])([i+Math.random()*10,j+Math.random()*10,z])(a))
	};
};